### PR TITLE
Added downward movement for Vertical Rails

### PR DESCRIPTION
**Pull request by EpyonProjects**

### CODE REVIEW

File: gm4_vertical_rails/data/gm4_vertical_rails/functions/rail/check_direction.mcfunction [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Checks if Vertical Rails go up or downward, prioritising up
+# @s = minecart, touching a Vertical Rail (ladder)
+# at @s
+# run from process
+
+tag @s add gm4_vert_rail_momentum
+
+execute as @s[tag=!gm4_vert_rail_down] at @s if block ~ ~1 ~ ladder run tag @s add gm4_vert_rail_up
```
Comment (Bloo-dev) [inline gm4_vertical_rails/data/gm4_vertical_rails/functions/rail/check_direction.mcfunction:8]:
```suggestion
execute if entity @s[tag=!gm4_vert_rail_down] if block ~ ~1 ~ ladder run tag @s add gm4_vert_rail_up
```
The function is already being called `as @s at @s` and no changes to the position of `@s` are made during the function, therefore `at @s` may be omitted to make this function run faster. Usage of `if entity @s` will also speed up the function in comparison to reselecting the frame of reference to `as @s`.

File: gm4_vertical_rails/data/gm4_vertical_rails/functions/rail/check_direction.mcfunction [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+execute as @s[tag=!gm4_vert_rail_up] at @s if block ~ ~-1 ~ ladder unless block ~ ~1 ~ ladder run tag @s add gm4_vert_rail_down
```
Comment (Bloo-dev) [inline gm4_vertical_rails/data/gm4_vertical_rails/functions/rail/check_direction.mcfunction:9]:
```suggestion
execute if entity @s[tag=!gm4_vert_rail_up] if block ~ ~-1 ~ ladder unless block ~ ~1 ~ ladder run tag @s add gm4_vert_rail_down
```
The function is already being called `at @s` and no changes to the position of `@s` are made during the function, therfore `at @s` may be ommited to make this function run faster. Usage of `if entity @s` will also speed up the function in comparison to reselecting the frame of reference to `as @s`.

File: gm4_vertical_rails/data/gm4_vertical_rails/functions/rail/apply_direction.mcfunction [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+# @s = minecart, touching a Vertical Rail (ladder), with gm4_vert_rails_momentum tag
+# at @s
+# run from process
+
+# if going up a Vertical Rail
+execute as @s[tag=gm4_vert_rail_up] at @s run function gm4_vertical_rails:rail/upward/momentum
+
+# if going down a Vertical Rail, or at bottom of Rail
+execute as @s[tag=gm4_vert_rail_down] at @s if block ~ ~-1 ~ ladder unless block ~ ~1 ~ ladder run function gm4_vertical_rails:rail/downward/momentum
+execute as @s[tag=gm4_vert_rail_down] at @s unless block ~ ~-1 ~ ladder run function gm4_vertical_rails:rail/downward/detach_ladder
```
Comment (Bloo-dev) [inline gm4_vertical_rails/data/gm4_vertical_rails/functions/rail/apply_direction.mcfunction:11]: The amount of reference frame changes in this function can be reduced to 0. See my comments attached to `check_direction.mcfunction`.
The momentum changes shouldn't alter the carts position until the next tick, so re-calling `at @s` shouldn't be necessary.

File: gm4_vertical_rails/data/gm4_vertical_rails/functions/process.mcfunction [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# tag minecarts upon attaching to a Vertical Rail
+execute as @s[tag=!gm4_vert_rail_momentum] at @s if block ~ ~ ~ ladder run function gm4_vertical_rails:rail/check_direction
+
+# while on a Vertical Rail
+execute as @s[tag=gm4_vert_rail_momentum] at @s if block ~ ~ ~ ladder run function gm4_vertical_rails:rail/apply_direction
+
+# once off a Vertical Rail
+execute as @s at @s unless block ~ ~ ~ ladder run function gm4_vertical_rails:rail/clear_tags
```
Comment (Bloo-dev) [inline gm4_vertical_rails/data/gm4_vertical_rails/functions/process.mcfunction:8]: Reduce the amount of reference frame changes; View my comment on `check_direction.mcfunction`.

**Review by Bloo-dev — CHANGES_REQUESTED** [review]

**Comment by EpyonProjects** [comment]
Dropping this :')

**Pull request closed without merging**